src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.hpp

```diff
@@ -359,42 +369,155 @@ class TableRelativePermeabilityHysteresis : public RelativePermeabilityBase
+
+    static constexpr char const * drainageNonWettingIntermediateRelPermTableNamesString()
+    { return "drainageNonWettingIntermediateRelPermTableNames"; }
+
+    static constexpr char const * imbibitionWettingRelPermTableNameString()
+    { return "imbibitionWettingRelPermTableName"; }
+
+    static constexpr char const * imbibitionNonWettingRelPermTableNameString()
+    { return "imbibitionNonWettingRelPermTableName"; }
 
   };
 
+  arrayView1d< real64 const > getPhaseMinVolumeFraction() const override
+  { return m_imbibitionPhaseMinVolFraction; };
+
+
+  /// for use in RelpermDriver to browse the drainage curves
+  /// by setting the MaxHistoricalNonWettingSat to Snwmin and MinWettingSat to Sw
+  GEOSX_HOST_DEVICE
+  std::tuple< integer, integer, integer > getPhaseIndex() const
+  {
+    using PT = PhaseType;
+    integer const ipWater = m_phaseOrder[PT::WATER];
+    integer const ipOil = m_phaseOrder[PT::OIL];
+    integer const ipGas = m_phaseOrder[PT::GAS];
+
+    integer ipWetting = -1, ipNonWetting = -1, ipInter = -1;
+
+    if( ipWater >= 0 && ipOil >= 0 && ipGas >= 0 )
+    {
+      ipWetting = ipWater;
+      ipInter = ipOil;
+      ipNonWetting = ipGas;
+    }
+    else if( ipWater < 0 )
+    {
+      ipWetting = ipOil;
+      ipNonWetting = ipGas;
+    }
+    else if( ipOil < 0 )
+    {
+      ipWetting = ipWater;
+      ipNonWetting = ipGas;
+    }
+    else if( ipGas < 0 )
+    {
+      ipWetting = ipWater;
+      ipNonWetting = ipOil;
+    }
+
+    //maybe a bit too pythonic
+    return std::make_tuple( ipWetting, ipInter, ipNonWetting );
+  }
+
+  GEOSX_HOST_DEVICE
+  void setMinMaxToDrainage( localIndex const k )
+  {
+
+    integer ipWetting, ipNonWetting, ipInter;
+    std::tie( ipWetting, ipInter, ipNonWetting ) = getPhaseIndex();
+
+    if( m_phaseHasHysteresis[ipNonWetting] )
+    {
+      m_phaseMaxHistoricalVolFraction[k][ipNonWetting] = m_drainagePhaseMaxVolFraction[ipNonWetting];
+      std::cout << " new Max NWet Historical " << m_phaseMaxHistoricalVolFraction[k][ipNonWetting] << std::endl;
+    }
+    if( m_phaseHasHysteresis[ipWetting] )
+    {
+      m_phaseMinHistoricalVolFraction[k][ipWetting] = m_drainagePhaseMinVolFraction[ipWetting];
+      std::cout << " new Min Wet Historical " << m_phaseMinHistoricalVolFraction[k][ipWetting] << std::endl;
+    }
+  }
+
+  /// for use in RelpermDriver to browse the drainage curves
+  /// by setting the MaxNonWettingSat to Snwmin and MinWettingSat to Sw
+  GEOSX_HOST_DEVICE
+  void setMinMaxToImbibition( localIndex const k )
+  {
+
+    integer ipWetting, ipNonWetting, ipInter;
+    std::tie( ipWetting, ipInter, ipNonWetting ) = getPhaseIndex();
+
+    if( m_phaseHasHysteresis[ipNonWetting] )
+    {
+      m_phaseMaxHistoricalVolFraction[k][ipNonWetting] = m_drainagePhaseMinVolFraction[ipNonWetting];
+    }
+    if( m_phaseHasHysteresis[ipWetting] )
+    {
+      m_phaseMinHistoricalVolFraction[k][ipWetting] = m_drainagePhaseMaxVolFraction[ipWetting];
+    }
+  }
```

Comment:
In the driver, I am going to propose another way to proceed that does not require the addition of these two functions `setMinMaxToImbibition` and `setMinMaxToDrainage` in the kernel wrapper. I am afraid that someone could misuse these functions from the solvers, which is possible in your implementation. Instead, we could try to do this directly in the `RelpermDriver` to avoid the need for these new functions here.

Reply:
I made those accepting pointer to `RelpermDriver` and the corresponding methods in `RelpermDriver` taking const and non-const Views to set the value properly in a visitor-like fashion. Do you think it would be safer ? All methods remaining public but the interface in `TableRelativePermeabilityHysteresis` being explicit in not using that in any other context than `RelpermDriver` ? 
It comes at the cost of a foward declaration either there https://github.com/GEOSX/GEOSX/blob/993caf706f8b5044f287e17b36104e40bd5ce3cb/src/coreComponents/constitutive/relativePermeability/RelpermDriver.hpp#L23-L27 
or here
https://github.com/GEOSX/GEOSX/blob/3c94b007d1ea8e02ef02ab7367a5c1545bffb82d/src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.hpp#L32-L33
and some more interdependence between those classes.